### ISSUE TITLE
Is value head gradient cut off?

### ISSUE THREAD

**Issue by likenneth** [lvwerra/trl#299]
I'm using this codebase to learn RLHF from raw. Looks there isn't a detailed enough tutorial out there for me. 

I can see the model boils down to a stack of Transformer encoder with two linear heads on it: a logit head, and a value head. During normal LLM pretraining, logit head is the only source of gradient for the many Transformer encoder blocks below, but what's happening now, in RLHF? I suppose logit head gradient is still there. How about the gradient from value head? 

I cannot find any code that explicitly cut it off, just to double check!

Looking forward to your answers!

**Comment by lvwerra** [comment]
Indeed, we do not cut it off and train both in parallel meaning that both the gradients from the LM head and the value head propagate back.

**Comment by likenneth** [comment]
Ahh thanks! That makes the most sense. I also wonder which paper(s) this implementation is based on, especially on technical details.

**Comment by lvwerra** [comment]
Sure, the implementation is from "Fine-Tuning Language Models from Human Preferences" by D. Ziegler et al. [[paper](https://arxiv.org/pdf/1909.08593.pdf), [code](https://github.com/openai/lm-human-preferences)].
